python/ray/_private/runtime_env/pip.py

```diff
@@ -0,0 +1,112 @@
+    if exit_code != 0:
+        shutil.rmtree(target_dir)
+        raise RuntimeError(f"Failed to install pip requirements:\n{output}")
+
+
+def get_uri(runtime_env: Dict) -> Optional[str]:
+    """Return `"pip://<hashed_dependencies>"`, or None if no GC required."""
+    pip = runtime_env.get("pip")
+    if pip is not None:
+        if isinstance(pip, list):
+            uri = "pip://" + _get_pip_hash(pip_list=pip)
+        else:
+            raise TypeError("pip field received by RuntimeEnvAgent must be "
+                            f"list, not {type(pip).__name__}.")
+    else:
+        uri = None
+    return uri
+
+
+class PipManager:
+    def __init__(self, resources_dir: str):
+        self._resources_dir = os.path.join(resources_dir, "pip")
+        try_to_create_directory(self._resources_dir)
+
+    def _get_path_from_hash(self, hash: str) -> str:
+        """Generate a path from the hash of a pip spec.
+
+        Example output:
+            /tmp/ray/session_2021-11-03_16-33-59_356303_41018/runtime_resources
+                /pip/ray-9a7972c3a75f55e976e620484f58410c920db091
+        """
+        return os.path.join(self._resources_dir, hash)
+
+    def delete_uri(self,
+                   uri: str,
+                   logger: Optional[logging.Logger] = default_logger) -> bool:
+        logger.info(f"Got request to delete URI {uri}")
+        protocol, hash = parse_uri(uri)
+        if protocol != Protocol.PIP:
+            raise ValueError("PipManager can only delete URIs with protocol "
+                             f"pip. Received protocol {protocol}, URI {uri}")
+
+        pip_env_path = self._get_path_from_hash(hash)
+        try:
+            shutil.rmtree(pip_env_path)
+            successful = True
+        except OSError:
+            successful = False
+            logger.warning(f"Error when deleting pip env {pip_env_path}.")
+        return successful
+
+    def setup(self,
+              runtime_env: RuntimeEnv,
+              context: RuntimeEnvContext,
+              logger: Optional[logging.Logger] = default_logger):
+        if not runtime_env.has_pip():
+            return
+
+        logger.debug(f"Setting up pip for runtime_env: {runtime_env}")
+        pip_packages: List[str] = runtime_env.pip_packages()
+        target_dir = self._get_path_from_hash(_get_pip_hash(pip_packages))
+
+        _install_pip_list_to_dir(pip_packages, target_dir, logger=logger)
+
+        # Despite Ray being removed from the input pip list during validation,
+        # other listed packages may include Ray as a dependency and therefore
+        # we may have inadvertently installed Ray in the target_dir anyway.
+        # Uninstall Ray here to make the workers use the Ray that is already
+        # installed in the cluster.
+        if os.environ.get(RAY_RUNTIME_ENV_ALLOW_RAY_IN_PIP) != 1:
+            ray_path = Path(target_dir) / "ray"
+            if ray_path.exists() and ray_path.is_dir():
+                shutil.rmtree(ray_path)
```

Comment:
oh this part looks pretty weird to me. if the user wanted to install Ray, let's just leave it? otherwise I'm not sure what this environment variable would even do

Reply:
Ah sorry, I worded the comment confusingly, let me improve it.  The deletion logic isn't for when the user wants to install ray, it's for when they put something like `ray_xgboost` in their pip list, which has `ray>=1.6` in its dependencies.  When we install `ray_xgboost`, it'll install the latest Ray into the target directory, and that will be prioritized over whatever Ray was already installed on the cluster.  So if the user is running anything other than the latest Ray, it'll fail with a version mismatch error.  

The environment variable is an escape hatch from the deletion logic, for when the user does for some reason want to include `ray` in their pip list.

Reply:
Oh wow, this is a pretty tricky edge case. New comment LGTM.